### PR TITLE
Diagnose bad 'ref' locals in switch sections

### LINKED ISSUE #44074: Process terminated. InvalidCastException: Unable to cast object of type 'LocalWithInitializer' to type 'SynthesizedLocal'.
**Version Used**: 
C:\Program Files\dotnet\sdk\3.1.300-preview-015135
VS Preview: Version 16.6.0 Preview 6.0 and 5.0
C# Tools   3.6.0-4.20251.5+910223b64f108fcf039012e0849befb46ace6e66

**Steps to Reproduce**:

I could not reproduce in a small code file.  There is exactly one line with which I can trigger this error at will. There is a detailed stack trace and a descriptive error. The full error is:
* Process terminated. System.InvalidCastException: Unable to cast object of type 'LocalWithInitializer' to type 'Microsoft.CodeAnalysis.CSharp.Symbols.SynthesizedLocal'.

```
public struct S {
    ...
    public IEnumerable<string> GetStrings() {
        ... (some nested switch statements)
        // this is the offending line:
        ref readonly var r= ref GetR(in this);
    }
    // Implementation of GetR; both R and P are structs wrapping a byte
    private ref readonly R GetR(in S @this) => ref Unsafe.As<P, R>(ref @this.P);
}
```

One oddity is that I had the offending line in several switch statements although the resulting variable `r` was not used.

**Expected Behavior**:
The code compiles.

**Actual Behavior**:
The process crashes.

Visual Studio shows in the Build Output:
```
3>C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\MSBuild\Current\Bin\Roslyn\Microsoft.CSharp.Core.targets(59,5): error MSB6006: "csc.exe" exited with code -2146232797.
```
I am lead to believe -2146232797 (0x80131623) is COR_E_FAILFAST due to #5348.

[CscCrashStackTrace.txt](https://github.com/dotnet/roslyn/files/4596942/CscCrashStackTrace.txt)

### PR DESCRIPTION
**Pull request by RikkiGibson**
Fixes #44074 by properly diagnosing the underlying issue.

There's maybe multiple ways to handle this. For instance we could find a spot where we gather up the locals in a switch section and check IsDirectlyInIterator there. We could find other spots where there are locals declared but not directly in a block or in a switch case and patch it all up. But it feels simpler and more likely to be right to just check at the time we bind the local whether it is a 'ref' local in an iterator.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 1]
```diff
@@ -1084,6 +1084,11 @@ protected BoundExpression BindInferredVariableInitializer(DiagnosticBag diagnost
 
                 if (localSymbol.RefKind != RefKind.None)
                 {
+                    if (IsDirectlyInIterator)
+                    {
+                        diagnostics.Add(ErrorCode.ERR_BadIteratorLocalType, localSymbol.Locations[0]);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:1089]: >ERR_BadIteratorLocalType [](start = 50, length = 24)

We used to report this error regardless whether there is an initializer or not.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:1089]: Also, consider bundling this check with the check for ERR_BadAsyncLocalType above. And sharing the code with BindForEachPartsWorker that has similar logic. BTW, why do we need the checks performed in BindForEachPartsWorker? Are we not calling this method for the locals checked there?

---
In reply to: [424592982](https://github.com/dotnet/roslyn/pull/44194#discussion_r424592982) [](ancestors = 424592982)

**Comment by AlekseyTs** [comment]
Where did we crash without the fix?

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 2)

**Comment by RikkiGibson** [comment]
> Where did we crash without the fix?

Stack trace is given in the linked issue, but here it is from running the test scenario in an old compiler:
```
Message: System.InvalidCastException: Unable to cast object of type 'LocalWithInitializer' to type 'Microsoft.CodeAnalysis.CSharp.Symbols.SynthesizedLocal'.
   at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.VisitAssignmentOperator(BoundAssignmentOperator node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriterWithStackGuard.VisitExpressionWithoutStackGuard(BoundExpression node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.VisitExpressionWithStackGuard(BoundExpression node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.VisitExpressionWithStackGuard(Int32& recursionDepth, BoundExpression node)
   at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.VisitExpressionStatement(BoundExpressionStatement node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriterWithStackGuard.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriter.VisitSequencePointWithSpan(BoundSequencePointWithSpan node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriterWithStackGuard.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriter.DoVisitList[T](ImmutableArray`1 list)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriter.VisitStatementList(BoundStatementList node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriterWithStackGuard.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriter.DoVisitList[T](ImmutableArray`1 list)
   at Microsoft.CodeAnalysis.CSharp.Symbols.MethodToClassRewriter.VisitBlock(BoundBlock node)
   at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.<>c__DisplayClass42_0.<VisitBlock>b__0()
   at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.PossibleIteratorScope(ImmutableArray`1 locals, Func`1 wrapped)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriterWithStackGuard.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriter.VisitSequencePointWithSpan(BoundSequencePointWithSpan node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriterWithStackGuard.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriter.DoVisitList[T](ImmutableArray`1 list)
   at Microsoft.CodeAnalysis.CSharp.Symbols.MethodToClassRewriter.VisitBlock(BoundBlock node)
   at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.<>c__DisplayClass42_0.<VisitBlock>b__0()
   at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.PossibleIteratorScope(ImmutableArray`1 locals, Func`1 wrapped)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriterWithStackGuard.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.MethodToStateMachineRewriter.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.IteratorMethodToStateMachineRewriter.GenerateMoveNextAndDispose(BoundStatement body, SynthesizedImplementationMethod moveNextMethod, SynthesizedImplementationMethod disposeMethod)
   at Microsoft.CodeAnalysis.CSharp.IteratorRewriter.GenerateEnumeratorImplementation()
   at Microsoft.CodeAnalysis.CSharp.IteratorRewriter.GenerateMethodImplementations()
   at Microsoft.CodeAnalysis.CSharp.StateMachineRewriter.Rewrite()
   at Microsoft.CodeAnalysis.CSharp.IteratorRewriter.Rewrite(BoundStatement body, MethodSymbol method, Int32 methodOrdinal, VariableSlotAllocator slotAllocatorOpt, TypeCompilationState compilationState, DiagnosticBag diagnostics, IteratorStateMachine& stateMachineType)
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.LowerBodyOrInitializer(MethodSymbol method, Int32 methodOrdinal, BoundStatement body, SynthesizedSubmissionFields previousSubmissionFields, TypeCompilationState compilationState, Boolean instrumentForDynamicAnalysis, DebugDocumentProvider debugDocumentProvider, ImmutableArray`1& dynamicAnalysisSpans, DiagnosticBag diagnostics, VariableSlotAllocator& lazyVariableSlotAllocator, ArrayBuilder`1 lambdaDebugInfoBuilder, ArrayBuilder`1 closureDebugInfoBuilder, StateMachineTypeSymbol& stateMachineTypeOpt)
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileMethod(MethodSymbol methodSymbol, Int32 methodOrdinal, ProcessedFieldInitializers& processedInitializers, SynthesizedSubmissionFields previousSubmissionFields, TypeCompilationState compilationState)
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileNamedType(NamedTypeSymbol containingType)
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.<>c__DisplayClass22_0.<CompileNamedTypeAsync>b__0()
```

Here we Debug.Assert that a ref local is synthesized before casting it--we do this because we're supposed to disallow the user from declaring ref locals in async methods and in iterator methods.

https://github.com/dotnet/roslyn/blob/91571a3bb038e05e7bf2ab87510273a1017faed0/src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/MethodToStateMachineRewriter.cs#L783

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs (hunk 1 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:1089]: > We used to report this error regardless whether there is an initializer or not.
>
> Also, consider bundling this check with the check for ERR_BadAsyncLocalType above. And sharing the code with BindForEachPartsWorker that has similar logic.

Thanks, will extract out a method which gives diagnostics for 'ref' variables in async or iterator methods.

> BTW, why do we need the checks performed in BindForEachPartsWorker? Are we not calling this method for the locals checked there?

We do not call BindVariableDeclaration for a foreach iteration variable.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 2]
```diff
@@ -1084,6 +1084,11 @@ protected BoundExpression BindInferredVariableInitializer(DiagnosticBag diagnost
 
                 if (localSymbol.RefKind != RefKind.None)
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:1085]: >if [](start = 16, length = 2)

Two nested if statements could be `&&`.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturnsTests.cs [hunk 3]
```diff
@@ -1184,6 +1184,37 @@ IEnumerable<int> M()
                 Diagnostic(ErrorCode.ERR_BadIteratorLocalType, "x").WithLocation(7, 26));
         }
 
+        [Fact]
+        public void RefReadonlyInSwitchCaseInIterator()
```
Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturnsTests.cs:1188]: >RefReadonlyInSwitchCaseInIterator [](start = 20, length = 33)

The LDM had discussed relaxing the requirements for embedded statements to permit a local declaration as an embedded statement. Please check that this fix also diagnoses a case like the following to ensure we don't miss it later

```csharp
    public IEnumerable<int> M1() {
        yield return 0;
        int y = 0;
        if (true)
            ref int x = ref y;
    }
```

**Comment by gafter** [comment]
Done reviewing commit 2.

**Pull request merged**